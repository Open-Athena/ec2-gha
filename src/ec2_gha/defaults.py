"""Default values for ec2-gha configuration."""

# Instance lifetime and timing defaults
MAX_INSTANCE_LIFETIME = "360"  # 6 hours (in minutes)
RUNNER_GRACE_PERIOD = "60"     # 1 minute (in seconds)
RUNNER_INITIAL_GRACE_PERIOD = "180"  # 3 minutes (in seconds)
RUNNER_POLL_INTERVAL = "10"    # 10 seconds
RUNNER_REGISTRATION_TIMEOUT = "300"  # 5 minutes (in seconds)

# EC2 instance defaults
EC2_INSTANCE_TYPE = "t3.medium"

# Instance naming default template
INSTANCE_NAME = "$repo/$name#$run"

# Default instance count
INSTANCE_COUNT = 1

# Home directory auto-detection sentinel
AUTO = "AUTO"
