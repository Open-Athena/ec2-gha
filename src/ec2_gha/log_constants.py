"""
Constants for log messages used across ec2-gha components.

These constants ensure consistency between the runner scripts that generate logs
and the analysis tools that parse them.
"""

# Log stream names (relative to instance ID prefix)
LOG_STREAM_RUNNER_SETUP = "runner-setup"
LOG_STREAM_JOB_STARTED = "job-started"
LOG_STREAM_JOB_COMPLETED = "job-completed"
LOG_STREAM_TERMINATION = "termination"
LOG_STREAM_RUNNER_DIAG = "runner-diag"

# Log message prefixes
LOG_PREFIX_JOB_STARTED = "Job started:"
LOG_PREFIX_JOB_COMPLETED = "Job completed:"

# Termination messages
LOG_MSG_TERMINATION_PROCEEDING = "proceeding with termination"
LOG_MSG_RUNNER_REMOVED = "Runner removed from GitHub successfully"

# Default CloudWatch log group
DEFAULT_CLOUDWATCH_LOG_GROUP = "/aws/ec2/github-runners"
