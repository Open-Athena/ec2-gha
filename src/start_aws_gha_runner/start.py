import importlib.resources
from dataclasses import dataclass, field
from string import Template

import boto3
from gha_runner import gh
from gha_runner.clouddeployment import CreateCloudInstance


@dataclass
class StartAWS(CreateCloudInstance):
    image_id: str
    instance_type: str
    home_dir: str
    repo: str
    region_name: str
    runner_release: str = ""
    tags: list[dict[str, str]] = field(default_factory=list)
    gh_runner_tokens: list[str] = field(default_factory=list)
    labels: str = ""
    subnet_id: str = ""
    security_group_id: str = ""
    iam_role: str = ""
    script: str = ""

    def _build_aws_params(self, user_data_params: dict) -> dict:
        """Build the parameters for the AWS API call.

        Parameters
        ----------
        count : int
            The number of instances to create.
        user_data_params : dict
            A dictionary of parameters to pass to the user

        Returns
        -------
        dict
            A dictionary of parameters for the AWS API call.

        """
        params = {
            "ImageId": self.image_id,
            "InstanceType": self.instance_type,
            "MinCount": 1,
            "MaxCount": 1,
            "UserData": self._build_user_data(**user_data_params),
        }
        if self.subnet_id != "":
            params["SubnetId"] = self.subnet_id
        if self.security_group_id != "":
            params["SecurityGroupIds"] = [self.security_group_id]
        if self.iam_role != "":
            params["IamInstanceProfile"] = {"Name": self.iam_role}
        if len(self.tags) > 0:
            specs = {"ResourceType": "instance", "Tags": self.tags}
            params["TagSpecifications"] = [specs]

        return params

    def _build_user_data(self, **kwargs) -> str:
        """Build the user data script.

        Parameters
        ----------
        kwargs : dict
            A dictionary of parameters to pass to the template.

        Returns
        -------
        str
            The user data script as a string.

        """
        template = importlib.resources.files("gha_runner").joinpath(
            "templates/user-script.sh.templ"
        )
        with template.open() as f:
            template = f.read()
            try:
                parsed = Template(template)
                return parsed.substitute(**kwargs)
            except Exception as e:
                raise Exception(f"Error parsing user data template: {e}")

    def create_instances(self) -> dict[str, str]:
        if not self.gh_runner_tokens:
            raise ValueError(
                "No GitHub runner tokens provided, cannot create instances."
            )
        if not self.runner_release:
            raise ValueError(
                "No runner release provided, cannot create instances."
            )
        if not self.home_dir:
            raise ValueError(
                "No home directory provided, cannot create instances."
            )
        if not self.image_id:
            raise ValueError("No image ID provided, cannot create instances.")
        if not self.instance_type:
            raise ValueError(
                "No instance type provided, cannot create instances."
            )
        ec2 = boto3.client("ec2", region_name=self.region_name)
        id_dict = {}
        for token in self.gh_runner_tokens:
            label = gh.GitHubInstance.generate_random_label()
            labels = self.labels
            if labels == "":
                labels = label
            else:
                labels = self.labels + "," + label
            user_data_params = {
                "token": token,
                "repo": self.repo,
                "homedir": self.home_dir,
                "script": self.script,
                "runner_release": self.runner_release,
                "labels": labels,
            }
            params = self._build_aws_params(user_data_params)
            result = ec2.run_instances(**params)
            instances = result["Instances"]
            id = instances[0]["InstanceId"]
            id_dict[id] = label
        return id_dict

    def wait_until_ready(self, ids: list[str], **kwargs):
        ec2 = boto3.client("ec2", self.region_name)
        waiter = ec2.get_waiter("instance_running")
        # Pass custom config for the waiter
        if kwargs:
            waiter.wait(InstanceIds=ids, WaiterConfig=kwargs)
        # Otherwise, use the default config
        else:
            waiter.wait(InstanceIds=ids)
