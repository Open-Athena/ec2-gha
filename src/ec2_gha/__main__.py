from ec2_gha.start import StartAWS
from gha_runner.gh import GitHubInstance
from gha_runner.clouddeployment import DeployInstance
from gha_runner.helper.input import EnvVarBuilder, check_required
import os


def main():
    env = dict(os.environ)
    required = ["GH_PAT", "AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY"]
    # Check that everything exists
    check_required(env, required)

    token = os.environ["GH_PAT"]
    # Make a copy of environment variables for immutability
    env = dict(os.environ)

    builder = (
        EnvVarBuilder(env)
        .update_state("INPUT_AWS_SUBNET_ID", "subnet_id")
        .update_state("INPUT_AWS_TAGS", "tags", is_json=True)
        .update_state("INPUT_EC2_HOME_DIR", "home_dir")
        .update_state("INPUT_EC2_IMAGE_ID", "image_id")
        .update_state("INPUT_EC2_INSTANCE_PROFILE", "iam_instance_profile")
        .update_state("INPUT_EC2_INSTANCE_TYPE", "instance_type")
        .update_state("INPUT_EC2_KEY_NAME", "key_name")
        .update_state("INPUT_EC2_ROOT_DEVICE_SIZE", "root_device_size", type_hint=int)
        .update_state("INPUT_EC2_SECURITY_GROUP_ID", "security_group_id")
        .update_state("INPUT_EC2_USERDATA", "userdata")
        .update_state("INPUT_EXTRA_GH_LABELS", "labels")
        .update_state("INPUT_INSTANCE_COUNT", "instance_count", type_hint=int)
        .update_state("INPUT_RUNNER_GRACE_PERIOD", "runner_grace_period")
        .update_state("INPUT_RUNNER_INITIAL_GRACE_PERIOD", "runner_initial_grace_period")
        .update_state("AWS_REGION", "region_name")        # default
        .update_state("INPUT_AWS_REGION", "region_name")  # input override
        .update_state("GITHUB_REPOSITORY", "repo")        # default
        .update_state("INPUT_REPO", "repo")               # input override
    )
    params = builder.params
    repo = params["repo"]
    # This needs to be handled here because the repo is required by the GitHub
    # instance
    if repo is None:
        raise Exception("Repo cannot be empty")

    # Instance count is not a keyword arg for StartAWS, so we remove it
    instance_count = params.pop("instance_count")

    gh = GitHubInstance(token=token, repo=repo)
    # This will create a new instance of StartAWS and configure it correctly
    deployment = DeployInstance(
        provider_type=StartAWS,
        cloud_params=params,
        gh=gh,
        count=instance_count,
    )
    # This will output the instance ids for using workflow syntax
    deployment.start_runner_instances()


if __name__ == "__main__":
    main()
