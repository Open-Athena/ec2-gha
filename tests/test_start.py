from unittest.mock import patch, mock_open, Mock

import pytest
from botocore.exceptions import WaiterError, ClientError
from moto import mock_aws

from ec2_gha.start import StartAWS


@pytest.fixture(scope="function")
def aws():
    with mock_aws():
        params = {
            "gh_runner_tokens": ["testing"],
            "home_dir": "/home/ec2-user",
            "image_id": "ami-0772db4c976d21e9b",
            "instance_type": "t2.micro",
            "region_name": "us-east-1",
            "repo": "omsf-eco-infra/awsinfratesting",
            "runner_grace_period": "120",
            "runner_release": "testing",
        }
        yield StartAWS(**params)


def test_build_user_data(aws, snapshot):
    """Test that template parameters are correctly substituted using snapshot testing"""
    params = {
        "cloudwatch_logs_group": "",  # Empty = disabled
        "github_run_id": "123456789",
        "github_run_number": "42",
        "github_workflow": "test-workflow",
        "homedir": "/home/test-user",
        "labels": "test-label",
        "max_instance_lifetime": "360",
        "repo": "test-org/test-repo",
        "runner_grace_period": "60",
        "runner_release": "https://example.com/runner.tar.gz",
        "script": "echo 'test script'",
        "ssh_pubkey": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC test@host",
        "token": "test-token-xyz",
        "userdata": "echo 'custom userdata'",
    }
    user_data = aws._build_user_data(**params)

    # Verify all substitutions happened (no template variables remain)
    template_vars = [ f'${k}' for k in params ]
    for var in template_vars:
        assert var not in user_data, f"Template variable {var} was not substituted"

    # Use snapshot to verify the entire output
    assert user_data == snapshot


def test_build_user_data_with_cloudwatch(aws, snapshot):
    """Test user data with CloudWatch Logs enabled using snapshot testing"""
    params = {
        "cloudwatch_logs_group": "/aws/ec2/github-runners",
        "github_run_id": "123456789",
        "github_run_number": "42",
        "github_workflow": "test-workflow",
        "homedir": "/home/test-user",
        "labels": "test-label",
        "max_instance_lifetime": "360",
        "repo": "test-org/test-repo",
        "runner_grace_period": "30",
        "runner_initial_grace_period": "120",
        "runner_release": "https://example.com/runner.tar.gz",
        "script": "echo 'test script'",
        "ssh_pubkey": "",
        "token": "test-token-xyz",
        "userdata": "",
    }
    user_data = aws._build_user_data(**params)

    # Verify all substitutions happened (no template variables remain)
    template_vars = [ f'${k}' for k in params ]
    for var in template_vars:
        assert var not in user_data, f"Template variable {var} was not substituted"

    # Use snapshot to verify the entire output
    assert user_data == snapshot


def test_build_user_data_missing_params(aws):
    """Test that missing required parameters raise an exception"""
    params = {
        "homedir": "/home/ec2-user",
        "repo": "omsf-eco-infra/awsinfratesting",
        "script": "echo 'Hello, World!'",
        "token": "test",
        "cloudwatch_logs_group": "",
        # Missing: labels, runner_release
    }
    with pytest.raises(Exception):
        aws._build_user_data(**params)


@pytest.fixture(scope="function")
def complete_params():
    params = {
        "gh_runner_tokens": ["test"],
        "home_dir": "/home/ec2-user",
        "iam_instance_profile": "test",
        "image_id": "ami-0772db4c976d21e9b",
        "instance_type": "t2.micro",
        "labels": "",
        "region_name": "us-east-1",
        "repo": "omsf-eco-infra/awsinfratesting",
        "root_device_size": 100,
        "runner_release": "test.tar.gz",
        "security_group_id": "test",
        "subnet_id": "test",
        "tags": [
            {"Key": "Name", "Value": "test"},
            {"Key": "Owner", "Value": "test"},
        ],
    }
    yield params


@patch.dict('os.environ', {
    'GITHUB_REPOSITORY': 'Open-Athena/ec2-gha',
    'GITHUB_WORKFLOW': 'CI',
    'GITHUB_SERVER_URL': 'https://github.com',
    'GITHUB_RUN_ID': '16725250800'
})
def test_build_aws_params(complete_params):
    user_data_params = {
        "cloudwatch_logs_group": "",
        "github_run_id": "16725250800",
        "github_run_number": "1",
        "github_workflow": "CI",
        "homedir": "/home/ec2-user",
        "labels": "label",
        "max_instance_lifetime": "360",
        "repo": "omsf-eco-infra/awsinfratesting",
        "runner_grace_period": "120",
        "runner_initial_grace_period": "180",
        "runner_release": "test.tar.gz",
        "script": "echo 'Hello, World!'",
        "ssh_pubkey": "",
        "token": "test",
        "userdata": "",
    }
    aws = StartAWS(**complete_params)
    params = aws._build_aws_params(user_data_params)

    # Test structure without checking exact UserData content
    assert params["ImageId"] == "ami-0772db4c976d21e9b"
    assert params["InstanceType"] == "t2.micro"
    assert params["MinCount"] == 1
    assert params["MaxCount"] == 1
    assert params["SubnetId"] == "test"
    assert params["SecurityGroupIds"] == ["test"]
    assert params["IamInstanceProfile"] == {"Name": "test"}
    assert params["InstanceInitiatedShutdownBehavior"] == "terminate"
    assert "UserData" in params
    assert params["TagSpecifications"] == [
        {
            "ResourceType": "instance",
            "Tags": [
                {"Key": "Name", "Value": "test"},
                {"Key": "Owner", "Value": "test"},
                {"Key": "repository", "Value": "Open-Athena/ec2-gha"},
                {"Key": "workflow", "Value": "CI"},
                {"Key": "gha_url", "Value": "https://github.com/Open-Athena/ec2-gha/actions/runs/16725250800"},
            ],
        }
    ]


def test_modify_root_disk_size(complete_params):
    mock_client = Mock()

    # Mock image data with all device mappings
    mock_image_data = {
        "Images": [{
            "RootDeviceName": "/dev/sda1",
            "BlockDeviceMappings": [
                {
                    "Ebs": {
                        "DeleteOnTermination": True,
                        "VolumeSize": 50,
                        "VolumeType": "gp3",
                        "Encrypted": False
                    },
                    "DeviceName": "/dev/sda1"
                },
                {
                    "DeviceName": "/dev/sdb",
                    "VirtualName": "ephemeral0"
                },
                {
                    "DeviceName": "/dev/sdc",
                    "VirtualName": "ephemeral1"
                }
            ]
        }]
    }

    def mock_describe_images(**kwargs):
        if kwargs.get('DryRun', False):
            raise ClientError(
                error_response={"Error": {"Code": "DryRunOperation"}},
                operation_name="DescribeImages"
            )
        return mock_image_data

    mock_client.describe_images = mock_describe_images
    aws = StartAWS(**complete_params)
    out = aws._modify_root_disk_size(mock_client, {})
    # Expected output should preserve all devices, only modifying root volume size
    expected_output = {
        "BlockDeviceMappings": [
            {
                "DeviceName": "/dev/sda1",
                "Ebs": {
                    "DeleteOnTermination": True,
                    "VolumeSize": 100,
                    "VolumeType": "gp3",
                    "Encrypted": False
                }
            },
            {
                "DeviceName": "/dev/sdb",
                "VirtualName": "ephemeral0"
            },
            {
                "DeviceName": "/dev/sdc",
                "VirtualName": "ephemeral1"
            }
        ]
    }
    assert out == expected_output


def test_modify_root_disk_size_permission_error(complete_params):
    mock_client = Mock()

    # Mock permission denied error
    mock_client.describe_images.side_effect = ClientError(
        error_response={'Error': {'Code': 'AccessDenied'}},
        operation_name='DescribeImages'
    )

    aws = StartAWS(**complete_params)

    with pytest.raises(ClientError) as exc_info:
        aws._modify_root_disk_size(mock_client, {})

    assert 'AccessDenied' in str(exc_info.value)


def test_modify_root_disk_size_no_change(complete_params):
    mock_client = Mock()
    complete_params["root_device_size"] = 0

    mock_image_data = {
        "Images": [{
            "RootDeviceName": "/dev/sda1",
            "BlockDeviceMappings": [
                {
                    "DeviceName": "/dev/sda1",
                    "Ebs": {
                        "VolumeSize": 50,
                        "VolumeType": "gp3"
                    }
                },
                {
                    "DeviceName": "/dev/sdb",
                    "VirtualName": "ephemeral0"
                }
            ]
        }]
    }

    def mock_describe_images(**kwargs):
        if kwargs.get('DryRun', False):
            raise ClientError(
                error_response={'Error': {'Code': 'DryRunOperation'}},
                operation_name='DescribeImages'
            )
        return mock_image_data

    mock_client.describe_images = mock_describe_images

    aws = StartAWS(**complete_params)
    input_params = {}
    result = aws._modify_root_disk_size(mock_client, input_params)

    # With root_device_size = 0, no modifications should be made
    assert result == input_params


def test_create_instance_with_labels(aws):
    aws.labels = "test"
    ids = aws.create_instances()
    assert len(ids) == 1


def test_create_instances(aws):
    ids = aws.create_instances()
    assert len(ids) == 1


def test_create_instances_missing_release(aws):
    aws.runner_release = ""
    with pytest.raises(
        ValueError, match="No runner release provided, cannot create instances."
    ):
        aws.create_instances()


def test_create_instances_missing_home_dir(aws):
    aws.home_dir = ""
    with pytest.raises(
        ValueError, match="No home directory provided, cannot create instances."
    ):
        aws.create_instances()


def test_create_instances_missing_tokens(aws):
    aws.gh_runner_tokens = []
    with pytest.raises(
        ValueError,
        match="No GitHub runner tokens provided, cannot create instances.",
    ):
        aws.create_instances()


def test_create_instances_missing_image_id(aws):
    aws.image_id = ""
    with pytest.raises(
        ValueError, match="No image ID provided, cannot create instances."
    ):
        aws.create_instances()


def test_create_instances_missing_instance_type(aws):
    aws.instance_type = ""
    with pytest.raises(
        ValueError, match="No instance type provided, cannot create instances."
    ):
        aws.create_instances()


def test_create_instances_missing_region_name(aws):
    aws.region_name = ""
    with pytest.raises(
        ValueError, match="No region name provided, cannot create instances."
    ):
        aws.create_instances()


def test_wait_until_ready(aws):
    ids = aws.create_instances()
    params = {
        "MaxAttempts": 1,
        "Delay": 5,
    }
    ids = list(ids)
    aws.wait_until_ready(ids, **params)


def test_wait_until_ready_dne(aws):
    # This is a fake instance id
    ids = ["i-xxxxxxxxxxxxxxxxx"]
    params = {
        "MaxAttempts": 1,
        "Delay": 5,
    }
    with pytest.raises(WaiterError):
        aws.wait_until_ready(ids, **params)


@pytest.mark.slow
def test_wait_until_ready_dne_long(aws):
    # This is a fake instance id
    ids = ["i-xxxxxxxxxxxxxxxxx"]
    # Runs with the default parameters
    with pytest.raises(WaiterError):
        aws.wait_until_ready(ids)


def test_set_instance_mapping(aws, monkeypatch):
    monkeypatch.setenv("GITHUB_OUTPUT", "mock_output_file")
    mapping = {"i-xxxxxxxxxxxxxxxxx": "test"}
    mock_file = mock_open()

    with patch("builtins.open", mock_file):
        aws.set_instance_mapping(mapping)

    # Should be called 4 times for single instance (mapping, instances, instance-id, label)
    assert mock_file.call_count == 4
    assert all(call[0][0] == "mock_output_file" for call in mock_file.call_args_list)


def test_set_instance_mapping_multiple(aws, monkeypatch):
    monkeypatch.setenv("GITHUB_OUTPUT", "mock_output_file")
    mapping = {"i-xxxxxxxxxxxxxxxxx": "test1", "i-yyyyyyyyyyyyyyyyy": "test2"}
    mock_file = mock_open()

    with patch("builtins.open", mock_file):
        aws.set_instance_mapping(mapping)

    # Should be called 2 times for multiple instances (mapping, instances only)
    assert mock_file.call_count == 2
    assert all(call[0][0] == "mock_output_file" for call in mock_file.call_args_list)
