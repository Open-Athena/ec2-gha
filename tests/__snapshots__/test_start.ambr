# serializer version: 1
# name: test_build_aws_params
  dict({
    'IamInstanceProfile': dict({
      'Name': 'test',
    }),
    'ImageId': 'ami-0772db4c976d21e9b',
    'InstanceInitiatedShutdownBehavior': 'terminate',
    'InstanceType': 't2.micro',
    'MaxCount': 1,
    'MinCount': 1,
    'SecurityGroupIds': list([
      'test',
    ]),
    'SubnetId': 'test',
    'TagSpecifications': list([
      dict({
        'ResourceType': 'instance',
        'Tags': list([
          dict({
            'Key': 'Name',
            'Value': 'test',
          }),
          dict({
            'Key': 'Owner',
            'Value': 'test',
          }),
          dict({
            'Key': 'Repository',
            'Value': 'Open-Athena/ec2-gha',
          }),
          dict({
            'Key': 'Workflow',
            'Value': 'CI',
          }),
          dict({
            'Key': 'URL',
            'Value': 'https://github.com/Open-Athena/ec2-gha/actions/runs/16725250800',
          }),
        ]),
      }),
    ]),
    'UserData': '''
      #!/bin/bash
      set -e
      
      # Essential variables from template substitution
      export debug=""
      export homedir="/home/ec2-user"
      export repo="omsf-eco-infra/awsinfratesting"
      export runner_tokens="test"
      export runner_labels="label"
      export cloudwatch_logs_group=""
      export runner_grace_period="61"
      export runner_initial_grace_period="181"
      export runner_poll_interval="11"
      export runner_registration_timeout="300"
      export max_instance_lifetime="360"
      export runners_per_instance="1"
      export runner_release="test.tar.gz"
      export ssh_pubkey="ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC test@host"
      export instance_name=""
      export action_sha="abc123def456789012345678901234567890abcd"
      
      # Custom userdata from user (if any)
      export userdata=""
      export script="echo 'Hello, World!'"
      
      # Log prefixes
      export log_prefix_job_started="Job started:"
      export log_prefix_job_completed="Job completed:"
      
      # Fetch and execute the main script from GitHub
      # action_sha has already been resolved from action_ref in Python for security and consistency
      SCRIPT_URL="https://raw.githubusercontent.com/Open-Athena/ec2-gha/${action_sha}/src/ec2_gha/scripts/runner-setup.sh"
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] Fetching main script from: $SCRIPT_URL (SHA: ${action_sha})" | tee -a /var/log/runner-setup.log
      
      # Try to download with retries
      for i in {1..5}; do
        if curl -sSL "$SCRIPT_URL" -o /tmp/runner-setup.sh; then
          echo "[$(date '+%Y-%m-%d %H:%M:%S')] Successfully downloaded runner setup script (attempt $i)" | tee -a /var/log/runner-setup.log
          break
        elif wget -q "$SCRIPT_URL" -O /tmp/runner-setup.sh; then
          echo "[$(date '+%Y-%m-%d %H:%M:%S')] Successfully downloaded runner setup script (attempt $i)" | tee -a /var/log/runner-setup.log
          break
        else
          echo "[$(date '+%Y-%m-%d %H:%M:%S')] Failed to download script (attempt $i), retrying..." | tee -a /var/log/runner-setup.log
          sleep 2
        fi
      
        if [ $i -eq 5 ]; then
          echo "[$(date '+%Y-%m-%d %H:%M:%S')] ERROR: Failed to download runner setup script after 5 attempts" | tee -a /var/log/runner-setup.log
          shutdown -h now
          exit 1
        fi
      done
      
      # Verify we got something
      if [ ! -s /tmp/runner-setup.sh ]; then
        echo "[$(date '+%Y-%m-%d %H:%M:%S')] ERROR: Downloaded script is empty" | tee -a /var/log/runner-setup.log
        shutdown -h now
        exit 1
      fi
      
      # Make it executable and run it
      chmod +x /tmp/runner-setup.sh
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] Executing runner setup script" | tee -a /var/log/runner-setup.log
      exec /tmp/runner-setup.sh
    ''',
  })
# ---
# name: test_build_aws_params_with_idx
  dict({
    'IamInstanceProfile': dict({
      'Name': 'test',
    }),
    'ImageId': 'ami-0772db4c976d21e9b',
    'InstanceInitiatedShutdownBehavior': 'terminate',
    'InstanceType': 't2.micro',
    'MaxCount': 1,
    'MinCount': 1,
    'SecurityGroupIds': list([
      'test',
    ]),
    'SubnetId': 'test',
    'TagSpecifications': list([
      dict({
        'ResourceType': 'instance',
        'Tags': list([
          dict({
            'Key': 'Name',
            'Value': 'ec2-gha/test#42 0',
          }),
          dict({
            'Key': 'Repository',
            'Value': 'Open-Athena/ec2-gha',
          }),
          dict({
            'Key': 'Workflow',
            'Value': 'CI',
          }),
          dict({
            'Key': 'URL',
            'Value': 'https://github.com/Open-Athena/ec2-gha/actions/runs/16725250800',
          }),
        ]),
      }),
    ]),
    'UserData': '''
      #!/bin/bash
      set -e
      
      # Essential variables from template substitution
      export debug=""
      export homedir="/home/ec2-user"
      export repo="omsf-eco-infra/awsinfratesting"
      export runner_tokens="test"
      export runner_labels="label"
      export cloudwatch_logs_group=""
      export runner_grace_period="61"
      export runner_initial_grace_period="181"
      export runner_poll_interval="11"
      export runner_registration_timeout="300"
      export max_instance_lifetime="360"
      export runners_per_instance="1"
      export runner_release="test.tar.gz"
      export ssh_pubkey="ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC test@host"
      export instance_name=""
      export action_sha="abc123def456789012345678901234567890abcd"
      
      # Custom userdata from user (if any)
      export userdata=""
      export script="echo 'Hello, World!'"
      
      # Log prefixes
      export log_prefix_job_started="Job started:"
      export log_prefix_job_completed="Job completed:"
      
      # Fetch and execute the main script from GitHub
      # action_sha has already been resolved from action_ref in Python for security and consistency
      SCRIPT_URL="https://raw.githubusercontent.com/Open-Athena/ec2-gha/${action_sha}/src/ec2_gha/scripts/runner-setup.sh"
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] Fetching main script from: $SCRIPT_URL (SHA: ${action_sha})" | tee -a /var/log/runner-setup.log
      
      # Try to download with retries
      for i in {1..5}; do
        if curl -sSL "$SCRIPT_URL" -o /tmp/runner-setup.sh; then
          echo "[$(date '+%Y-%m-%d %H:%M:%S')] Successfully downloaded runner setup script (attempt $i)" | tee -a /var/log/runner-setup.log
          break
        elif wget -q "$SCRIPT_URL" -O /tmp/runner-setup.sh; then
          echo "[$(date '+%Y-%m-%d %H:%M:%S')] Successfully downloaded runner setup script (attempt $i)" | tee -a /var/log/runner-setup.log
          break
        else
          echo "[$(date '+%Y-%m-%d %H:%M:%S')] Failed to download script (attempt $i), retrying..." | tee -a /var/log/runner-setup.log
          sleep 2
        fi
      
        if [ $i -eq 5 ]; then
          echo "[$(date '+%Y-%m-%d %H:%M:%S')] ERROR: Failed to download runner setup script after 5 attempts" | tee -a /var/log/runner-setup.log
          shutdown -h now
          exit 1
        fi
      done
      
      # Verify we got something
      if [ ! -s /tmp/runner-setup.sh ]; then
        echo "[$(date '+%Y-%m-%d %H:%M:%S')] ERROR: Downloaded script is empty" | tee -a /var/log/runner-setup.log
        shutdown -h now
        exit 1
      fi
      
      # Make it executable and run it
      chmod +x /tmp/runner-setup.sh
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] Executing runner setup script" | tee -a /var/log/runner-setup.log
      exec /tmp/runner-setup.sh
    ''',
  })
# ---
# name: test_build_user_data
  '''
  #!/bin/bash
  set -e
  
  # Essential variables from template substitution
  export debug=""
  export homedir="/home/ec2-user"
  export repo="omsf-eco-infra/awsinfratesting"
  export runner_tokens="test"
  export runner_labels="label"
  export cloudwatch_logs_group=""
  export runner_grace_period="61"
  export runner_initial_grace_period="181"
  export runner_poll_interval="11"
  export runner_registration_timeout="300"
  export max_instance_lifetime="360"
  export runners_per_instance="1"
  export runner_release="test.tar.gz"
  export ssh_pubkey="ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC test@host"
  export instance_name=""
  export action_sha="abc123def456789012345678901234567890abcd"
  
  # Custom userdata from user (if any)
  export userdata=""
  export script="echo 'Hello, World!'"
  
  # Log prefixes
  export log_prefix_job_started="Job started:"
  export log_prefix_job_completed="Job completed:"
  
  # Fetch and execute the main script from GitHub
  # action_sha has already been resolved from action_ref in Python for security and consistency
  SCRIPT_URL="https://raw.githubusercontent.com/Open-Athena/ec2-gha/${action_sha}/src/ec2_gha/scripts/runner-setup.sh"
  echo "[$(date '+%Y-%m-%d %H:%M:%S')] Fetching main script from: $SCRIPT_URL (SHA: ${action_sha})" | tee -a /var/log/runner-setup.log
  
  # Try to download with retries
  for i in {1..5}; do
    if curl -sSL "$SCRIPT_URL" -o /tmp/runner-setup.sh; then
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] Successfully downloaded runner setup script (attempt $i)" | tee -a /var/log/runner-setup.log
      break
    elif wget -q "$SCRIPT_URL" -O /tmp/runner-setup.sh; then
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] Successfully downloaded runner setup script (attempt $i)" | tee -a /var/log/runner-setup.log
      break
    else
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] Failed to download script (attempt $i), retrying..." | tee -a /var/log/runner-setup.log
      sleep 2
    fi
  
    if [ $i -eq 5 ]; then
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] ERROR: Failed to download runner setup script after 5 attempts" | tee -a /var/log/runner-setup.log
      shutdown -h now
      exit 1
    fi
  done
  
  # Verify we got something
  if [ ! -s /tmp/runner-setup.sh ]; then
    echo "[$(date '+%Y-%m-%d %H:%M:%S')] ERROR: Downloaded script is empty" | tee -a /var/log/runner-setup.log
    shutdown -h now
    exit 1
  fi
  
  # Make it executable and run it
  chmod +x /tmp/runner-setup.sh
  echo "[$(date '+%Y-%m-%d %H:%M:%S')] Executing runner setup script" | tee -a /var/log/runner-setup.log
  exec /tmp/runner-setup.sh
  '''
# ---
# name: test_build_user_data_with_cloudwatch
  '''
  #!/bin/bash
  set -e
  
  # Essential variables from template substitution
  export debug=""
  export homedir="/home/ec2-user"
  export repo="omsf-eco-infra/awsinfratesting"
  export runner_tokens="test"
  export runner_labels="label"
  export cloudwatch_logs_group="/aws/ec2/github-runners"
  export runner_grace_period="61"
  export runner_initial_grace_period="181"
  export runner_poll_interval="11"
  export runner_registration_timeout="300"
  export max_instance_lifetime="360"
  export runners_per_instance="1"
  export runner_release="test.tar.gz"
  export ssh_pubkey=""
  export instance_name=""
  export action_sha="abc123def456789012345678901234567890abcd"
  
  # Custom userdata from user (if any)
  export userdata=""
  export script="echo 'Hello, World!'"
  
  # Log prefixes
  export log_prefix_job_started="Job started:"
  export log_prefix_job_completed="Job completed:"
  
  # Fetch and execute the main script from GitHub
  # action_sha has already been resolved from action_ref in Python for security and consistency
  SCRIPT_URL="https://raw.githubusercontent.com/Open-Athena/ec2-gha/${action_sha}/src/ec2_gha/scripts/runner-setup.sh"
  echo "[$(date '+%Y-%m-%d %H:%M:%S')] Fetching main script from: $SCRIPT_URL (SHA: ${action_sha})" | tee -a /var/log/runner-setup.log
  
  # Try to download with retries
  for i in {1..5}; do
    if curl -sSL "$SCRIPT_URL" -o /tmp/runner-setup.sh; then
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] Successfully downloaded runner setup script (attempt $i)" | tee -a /var/log/runner-setup.log
      break
    elif wget -q "$SCRIPT_URL" -O /tmp/runner-setup.sh; then
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] Successfully downloaded runner setup script (attempt $i)" | tee -a /var/log/runner-setup.log
      break
    else
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] Failed to download script (attempt $i), retrying..." | tee -a /var/log/runner-setup.log
      sleep 2
    fi
  
    if [ $i -eq 5 ]; then
      echo "[$(date '+%Y-%m-%d %H:%M:%S')] ERROR: Failed to download runner setup script after 5 attempts" | tee -a /var/log/runner-setup.log
      shutdown -h now
      exit 1
    fi
  done
  
  # Verify we got something
  if [ ! -s /tmp/runner-setup.sh ]; then
    echo "[$(date '+%Y-%m-%d %H:%M:%S')] ERROR: Downloaded script is empty" | tee -a /var/log/runner-setup.log
    shutdown -h now
    exit 1
  fi
  
  # Make it executable and run it
  chmod +x /tmp/runner-setup.sh
  echo "[$(date '+%Y-%m-%d %H:%M:%S')] Executing runner setup script" | tee -a /var/log/runner-setup.log
  exec /tmp/runner-setup.sh
  '''
# ---
